pre_commit_hooks/check_twincat_versions.py

```diff
@@ -0,0 +1,143 @@
+    else:
+        version_pattern = r'(TcVersion="[^"]*")'
+        new_xml_content = re.sub(
+            version_pattern, r'\g<1> TcVersionFixed="' + new_value + r'"', xml_content
+        )
+
+    return new_xml_content
+
+
+def main(args=None):
+    if args is None:
+        parser = argparse.ArgumentParser()
+        parser.add_argument(
+            "filenames", nargs="+", help="List of XML filenames to process"
+        )
+        parser.add_argument(
+            "--target-version", type=str, help="Target TcVersion to enforce"
+        )
+        parser.add_argument(
+            "--fix",
+            action="store_true",
+            help="Fix the versions if they do not match the target version",
+        )
+        parser.add_argument(
+            "--reason", type=str, help="Reason for targeting a specific version"
+        )
+        parser.add_argument(
+            "--pinned",
+            action=argparse.BooleanOptionalAction,
+            help="Check if the TwinCAT versions should be pinned",
+        )
+
+        args = parser.parse_args()
+
+    try:
+        versions = {}
+        pinned = {}
+        for filename in args.filenames:
+            with open(filename, "r") as file:
+                xml_content = file.read()
+                versions[filename] = get_tc_version(xml_content)
+                pinned[filename] = tc_version_pinned(xml_content)
+
+        itemize = "\n -"
+        exception_message = ""
+        if args.target_version:
+            mismatched_files = [
+                fname for fname, ver in versions.items() if ver != args.target_version
+            ]
+            if mismatched_files:
+                reason_msg = f"\nReason: {args.reason}" if args.reason else ""
+                if args.fix:
+                    for filename in mismatched_files:
+                        with open(filename, "r") as file:
+                            xml_content = file.read()
+                        fixed_content = fix_tc_version(xml_content, args.target_version)
+                        with open(filename, "w") as file:
+                            file.write(fixed_content)
+
+                    print(
+                        f"Fixed TwinCAT versions for:{itemize}{itemize.join(mismatched_files)}{reason_msg}"
+                    )
+                else:
+                    exception_message += (
+                        "The following files are not set to the targeted TwinCAT version "
+                        f"{args.target_version}:{itemize}{itemize.join(mismatched_files)}{reason_msg}"
+                    )
+        else:
+            unique_versions = set(versions.values())
+            if len(unique_versions) > 1:
+                exception_message += (
+                    "Not all files have the same TwinCAT version:"
+                    f"{itemize}"
+                    + itemize.join(f"{fname}: {ver}" for fname, ver in versions.items())
+                )
+
+        if args.pinned:
```

Comment:
Small nitpick: this invalidates a potential `--no-pinned` option from the BooleanOptionalAction input.
```suggestion
        if args.pinned is not None:
```

Reply:
Good catch